File: Super-Resolution-in-OpenCV/super_res.py
#!/usr/bin/env python3
"""Upscale an image with OpenCV's DNN super-resolution module."""

from __future__ import annotations

import argparse
from pathlib import Path
import sys

import cv2


PROJECT_DIR = Path(__file__).resolve().parent
SUPPORTED_SCALES = {
    "edsr": {2, 3, 4},
    "espcn": {2, 3, 4},
    "fsrcnn": {2, 3, 4},
    "lapsrn": {2, 4, 8},
}


def validate_model_configuration(algorithm: str, scale: int) -> None:
    """Validate an algorithm/scale pair before loading a model."""
    valid_scales = SUPPORTED_SCALES.get(algorithm)
    if valid_scales is None:
        choices = ", ".join(sorted(SUPPORTED_SCALES))
        raise ValueError(f"Unsupported algorithm '{algorithm}'. Choose from: {choices}.")
    if scale not in valid_scales:
        choices = ", ".join(str(value) for value in sorted(valid_scales))
        raise ValueError(
            f"Scale x{scale} is not supported for {algorithm}. "
            f"Choose from: {choices}."
        )


def upscale_image(
    image,
    model_path: Path,
    algorithm: str = "espcn",
    scale: int = 4,
):
    """Return an image enlarged by the requested DNN super-resolution model."""
    if image is None or image.size == 0:
        raise ValueError("The input image is empty.")

    validate_model_configuration(algorithm, scale)
    model_path = Path(model_path)
    if not model_path.is_file():
        raise FileNotFoundError(
            f"Model not found: {model_path}. Run 'python download_model.py' "
            "or pass --model."
        )

    if not hasattr(cv2, "dnn_superres"):
        raise RuntimeError(
            "This OpenCV build does not include dnn_superres. "
            "Install opencv-contrib-python or build OpenCV with opencv_contrib."
        )

    super_resolver = cv2.dnn_superres.DnnSuperResImpl_create()
    super_resolver.readModel(str(model_path))
    super_resolver.setModel(algorithm, scale)
    result = super_resolver.upsample(image)

    expected_size = (image.shape[0] * scale, image.shape[1] * scale)
    if result.shape[:2] != expected_size:
        raise RuntimeError(
            "Unexpected output dimensions: "
            f"got {result.shape[1]}x{result.shape[0]}, "
            f"expected {expected_size[1]}x{expected_size[0]}."
        )
    return result


def parse_args(argv: list[str] | None = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Upscale an image with OpenCV DNN super resolution."
    )
    parser.add_argument(
        "--input",
        type=Path,
        default=PROJECT_DIR / "image.png",
        help="Input image (default: the bundled image.png).",
    )
    parser.add_argument(
        "--model",
        type=Path,
        default=PROJECT_DIR / "models" / "ESPCN_x4.pb",
        help="TensorFlow .pb model file.",
    )
    parser.add_argument(
        "--algorithm",
        choices=sorted(SUPPORTED_SCALES),
        default="espcn",
        help="Model architecture (default: espcn).",
    )
    parser.add_argument(
        "--scale",
        type=int,
        default=4,
        help="Upscaling factor encoded by the model (default: 4).",
    )
    parser.add_argument(
        "--output",
        type=Path,
        default=PROJECT_DIR / "output.png",
        help="Output image path.",
    )
    return parser.parse_args(argv)


def main(argv: list[str] | None = None) -> int:
    args = parse_args(argv)

    try:
        image = cv2.imread(str(args.input), cv2.IMREAD_COLOR)
        if image is None:
            raise FileNotFoundError(f"Input image not found or unreadable: {args.input}")

        result = upscale_image(
            image=image,
            model_path=args.model,
            algorithm=args.algorithm,
            scale=args.scale,
        )
        args.output.parent.mkdir(parents=True, exist_ok=True)
        if not cv2.imwrite(str(args.output), result):
            raise OSError(f"OpenCV could not write: {args.output}")
    except (FileNotFoundError, OSError, RuntimeError, ValueError, cv2.error) as error:
        print(f"Error: {error}", file=sys.stderr)
        return 2

    print(
        f"Upscaled {image.shape[1]}x{image.shape[0]} -> "
        f"{result.shape[1]}x{result.shape[0]} and wrote {args.output}"
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: Super-Resolution-in-OpenCV/super_res.cpp
#include <opencv2/dnn_superres.hpp>
#include <opencv2/imgcodecs.hpp>

#include <filesystem>
#include <iostream>
#include <string>

namespace fs = std::filesystem;

const char* kKeys =
    "{help h usage ? |      | Show this message }"
    "{input i        |image.png| Input image }"
    "{model m        |models/ESPCN_x4.pb| TensorFlow .pb model }"
    "{algorithm a    |espcn| Model architecture: edsr, espcn, fsrcnn, or lapsrn }"
    "{scale s        |4     | Upscaling factor encoded by the model }"
    "{output o       |output.png| Output image }";

bool isSupportedConfiguration(const std::string& algorithm, int scale) {
    if (algorithm == "lapsrn") {
        return scale == 2 || scale == 4 || scale == 8;
    }
    if (algorithm == "edsr" || algorithm == "espcn" ||
        algorithm == "fsrcnn") {
        return scale == 2 || scale == 3 || scale == 4;
    }
    return false;
}

int main(int argc, char** argv) {
    cv::CommandLineParser parser(argc, argv, kKeys);
    parser.about("OpenCV DNN super-resolution example");
    if (parser.has("help")) {
        parser.printMessage();
        return 0;
    }
    if (!parser.check()) {
        parser.printErrors();
        return 2;
    }

    const std::string input_path = parser.get<std::string>("input");
    const std::string model_path = parser.get<std::string>("model");
    const std::string algorithm = parser.get<std::string>("algorithm");
    const int scale = parser.get<int>("scale");
    const fs::path output_path = parser.get<std::string>("output");

    if (!isSupportedConfiguration(algorithm, scale)) {
        std::cerr << "Error: unsupported --algorithm/--scale combination. "
                     "Use edsr, espcn, or fsrcnn with x2/x3/x4, or lapsrn "
                     "with x2/x4/x8.\n";
        return 2;
    }
    if (!fs::is_regular_file(model_path)) {
        std::cerr << "Error: model not found: " << model_path
                  << ". Run 'python download_model.py' or pass --model.\n";
        return 2;
    }

    const cv::Mat image = cv::imread(input_path, cv::IMREAD_COLOR);
    if (image.empty()) {
        std::cerr << "Error: input image not found or unreadable: "
                  << input_path << '\n';
        return 2;
    }

    try {
        cv::dnn_superres::DnnSuperResImpl super_resolver;
        super_resolver.readModel(model_path);
        super_resolver.setModel(algorithm, scale);

        cv::Mat result;
        super_resolver.upsample(image, result);
        const cv::Size expected(image.cols * scale, image.rows * scale);
        if (result.size() != expected) {
            std::cerr << "Error: unexpected output dimensions: "
                      << result.cols << 'x' << result.rows << "; expected "
                      << expected.width << 'x' << expected.height << '\n';
            return 3;
        }

        if (output_path.has_parent_path()) {
            fs::create_directories(output_path.parent_path());
        }
        if (!cv::imwrite(output_path.string(), result)) {
            std::cerr << "Error: OpenCV could not write: " << output_path << '\n';
            return 3;
        }
        std::cout << "Upscaled " << image.cols << 'x' << image.rows << " -> "
                  << result.cols << 'x' << result.rows << " and wrote "
                  << output_path << '\n';
    } catch (const cv::Exception& error) {
        std::cerr << "OpenCV error: " << error.what() << '\n';
        return 3;
    } catch (const fs::filesystem_error& error) {
        std::cerr << "Filesystem error: " << error.what() << '\n';
        return 3;
    }

    return 0;
}
